### PR TITLE
Added and passed new tests for optional attendees

### PR DESCRIPTION
**Pull request by arianazhu**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -29,23 +29,38 @@
      * request: new meeting request
      */
     public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-        Collection<String> meeting_attendees = request.getAttendees();
-        if (meeting_attendees.size() <= 0) {
+        Collection<String> mandatory_attendees = request.getAttendees();
+        Collection<String> optional_attendees = request.getOptionalAttendees();
+        if ((mandatory_attendees.size() <= 0) && (optional_attendees.size() <= 0)) {
             return Arrays.asList(TimeRange.WHOLE_DAY);
```
Comment (taruss2011) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: This is more a of a style choice, but if you find that the indentation of the main code is getting bothersome, you can arrange to limit the indentation depth when control leaves from inside `if` statements by not using `else` or `else if` clauses.
```
if (...)
  return ...
if (...)
  return ...
rest-of-code
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -29,23 +29,38 @@
...
         }
         else if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
             return Arrays.asList();
         }
         else {
-            List<TimeRange> busy_times = getBusyTimes(events, meeting_attendees);
+            List<TimeRange> mandatory_busy_times = getBusyTimes(events, mandatory_attendees);
+            List<TimeRange> all_busy_times = getBusyTimes(events, optional_attendees);
```
Comment (taruss2011) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Is this **all** busy times or just those for the optional attendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -29,23 +29,38 @@
...
+        Collection<String> optional_attendees = request.getOptionalAttendees();
+        if ((mandatory_attendees.size() <= 0) && (optional_attendees.size() <= 0)) {
             return Arrays.asList(TimeRange.WHOLE_DAY);
         }
         else if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
             return Arrays.asList();
         }
         else {
-            List<TimeRange> busy_times = getBusyTimes(events, meeting_attendees);
+            List<TimeRange> mandatory_busy_times = getBusyTimes(events, mandatory_attendees);
```
Comment (taruss2011) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Optimization:
If there are any optional attendees, you should first handle the scheduling for all busy times and return that if there is a solution.
Only if there is no solution to that case, compute the scheduling solution for just mandatory attendees. That way you don't need to do this work if there is an answer to the more constrained case.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -29,23 +29,38 @@
...
+
+            List<TimeRange> mandatory_usable_times = getUsableTimes(mandatory_available_times, request.getDuration());
+            List<TimeRange> all_usable_times = getUsableTimes(all_available_times, request.getDuration());
+
+            if (all_usable_times.size() > 0) {
+                return all_usable_times;
+            }
+            else if ((optional_attendees.size() > 0) && (mandatory_attendees.size() <= 0)) {
+                return all_usable_times;
+            }
```
Comment (taruss2011) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Here is where you should do the handling of mandatory times, including extraction, sorting and getting usable times.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -81,11 +96,7 @@ else if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
...
             }
-            // Case 2: busy_start is right at free_start ?? figure out points
-            if (busy_start == free_start) {
-
-            }
-            // Case 3: end comes after start, and no busy events will overlap
+            // Case 2: end comes after start, and no busy events will overlap
             available_times.add(TimeRange.fromStartEnd(free_start, busy_start, false));
             free_start = busy_slot.end();
         }
```
Comment (taruss2011) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:103]: If at the end of the loop, `free_start` = `END_OF_DAY`, doesn't this add a zero-length interval to `available_times`.  And is there a similar issue if a busy time begins at `START_OF_DAY`?

Do you have test cases for the edge conditions of
```
Start          10am              2pm            End
|---------------|                 |--------------|
```
where you should get a single free time of 10am-2pm?

Actually, when I look at the test `optionalAttendeesOnlyAndAreAvailable`, it covers this case. Does that test case pass?
When looking at the code, I don't see how it would pass. It seems to me that you have the following things happening:
```
free_start = START_OF_DAY
loop
  busy_start = START_OF_DAY
  if (busy_start < free_start) // false
    ...
  available_times.add(free_start, busy_start)
```

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -270,5 +272,143 @@ public void notEnoughRoom() {
...
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void optionalAttendeesUnavailable() {
+    // Have each person have different events. We should see three options because each person has
+    // split the restricted times, and no time slots exists that both mandatory and optional
+    // attendees can attend. (based on everyAttendeeIsConsidered)
+    //
+    // Events  : |--------------C--------------|     <- only for optional attendee
```
Comment (taruss2011) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:282]: The diagrams in the test comments are great!

**Review by taruss2011 — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (arianazhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: all busy times, so taking every attendee into account

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (arianazhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:103]: Yes, the test case passes. I believe the getUsableTimes() function should remove any time range that isn't long enough for the meeting duration?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (taruss2011) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: All.
I should look on the next line, where the mandatory busy times get added to all busy times.  Oops.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (taruss2011) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:103]: Ah, that is correct.
I missed where that was called.

**Review by taruss2011 — APPROVED** [review]
Well, I guess I missed a couple of key lines when I reviewed this.
Sorry about that.
